### ISSUE TITLE
Packer hangs while running shell provisioners

### ISSUE THREAD

**Issue by aol-nnov** [hashicorp/packer#4627]
@mwhooker I suppose it has something to do with #2447 

My setup is as follows:
```
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.1 LTS
Release:	16.04
Codename:	xenial
$ packer version
Packer v0.12.3
```
I'm using @jakobadam 's temlates from https://github.com/jakobadam/packer-qemu-templates

Namely, json in question is https://github.com/jakobadam/packer-qemu-templates/blob/master/centos/centos-6.7-server-x86_64-vagrant.json
with minor modifications.

When it comes to provisioning, output hangs on the first script and then times out.

All this happens if I execute packer from hw console or inside of X session.
If I ssh into machine and then run packer (even if it is ```ssh user@localhost```) provisioning runs just fine.

Also, if I run packer as root, it also succeeds.

I'm 99% sure that it's a tty problem, but I'm not sure what to tweak in my system.

Any help is highly appreciated!

**Comment by mwhooker** [comment]
can you post Debug log output from `PACKER_LOG=1 packer build template.json`. Please paste this in a gist https://gist.github.com

**Comment by aol-nnov** [comment]
@mwhooker unfortunately it's late here and I do not have remote access to my work place.
here is a snippet from a chat with my colleague:
```
2017/03/03 14:45:16 [INFO] 325 bytes written for 'uploadData'
2017/03/03 14:45:16 packer: 2017/03/03 14:45:16 opening new ssh session
2017/03/03 14:45:16 packer: 2017/03/03 14:45:16 Starting remote scp process:  scp -vt /tmp
2017/03/03 14:45:16 packer: 2017/03/03 14:45:16 Started SCP session, beginning transfers...
2017/03/03 14:45:16 packer: 2017/03/03 14:45:16 Copying input data into temporary file so we can read the length
2017/03/03 14:45:16 packer: 2017/03/03 14:45:16 [DEBUG] scp: Uploading script_3565.sh: perms=C0644 size=325
2017/03/03 14:45:16 packer: 2017/03/03 14:45:16 SCP session complete, closing stdin pipe.
2017/03/03 14:45:16 packer: 2017/03/03 14:45:16 Waiting for SSH session to complete.
2017/03/03 14:45:16 packer: 2017/03/03 14:45:16 non-zero exit status: 255
2017/03/03 14:45:16 packer: 2017/03/03 14:45:16 Retryable error: Error uploading script: Process exited with status 255
```

During a quick check it came out that files are not transferred to vm at all.

I'll post full debug log to the gist tomorrow as soon as I get to work.

**Comment by mwhooker** [comment]
thanks!

**Comment by aol-nnov** [comment]
without ```PACKER_LOG=1``` it prints out following logs (a snippet from chat again, sorry):
```
==> mcbc5-rev8-server-vagrant: Typing the boot command over VNC...
==> mcbc5-rev8-server-vagrant: Waiting for SSH to become available...
==> mcbc5-rev8-server-vagrant: Connected to SSH!
==> mcbc5-rev8-server-vagrant: Provisioning with shell script: scripts/vagrant.sh
[huge delay here]
==> mcbc5-rev8-server-vagrant: Deleting output directory...
Build 'mcbc5-rev8-server-vagrant' errored: Retryable error: Error uploading script: Process exited with status 255

==> Some builds didn't complete successfully and had errors:
--> mcbc5-rev8-server-vagrant: Retryable error: Error uploading script: Process exited with status 255

==> Builds finished but no artifacts were created.
```
```scripts/vagrant.sh``` is the first script int my provisioning section and its contents is identical to [this file](https://github.com/jakobadam/packer-qemu-templates/blob/master/centos/scripts/vagrant.sh)

**Comment by aol-nnov** [comment]
Well, this is embarrassing! I can no longer reproduce it on my host. Banging my head two days in a row, no changes on host and it no longer fails when I need to provide logs..

Anyway, it still fails on colleague's host with exactly the same setup (ubuntu 16.04 server, packer v0.12.3)
https://gist.github.com/aol-nnov/72106d921b68d07d69fa454a08c958e3

packer.log was captured inside X session, packer-ssh.log was captured after ```ssh localhost``` inside that X session.

**Comment by harmishhk** [comment]
My few cents on this topic:
Recently I updated to 0.12.3 version and this problem reappeared (I only use VirtualBox builder). After few days of frustration I realized it might be due to the new version so I checked with older versions. And in fact, after downgrading to version 0.10.1 this problem disappears. (I also checked with version 0.11.0 where the problem still exist.)

It seems that the shell provisioner hangs when large amount of text is output from the running process. For example, in my use, it always hangs when my shell script executes `sudo apt-get -y install ubuntu-desktop` on Ubuntu 16.04 server.

**Comment by mwhooker** [comment]
I'm unable to reproduce this on ubuntu.

I had to rewrite the scripts and the json for the linked repository to make it use centos 6.8, since the mirror links for 6.7 were 404ing

If you can start the qemu build with gui attached, can you see if the kickstart file gets processed correctly? It looks like packer is having trouble uploading files to the guest. Make sure you can connect to it over ssh and upload files to it. 

@harmishhk Would you mind opening another ticket for this? I wasn't able to reproduce the issue, so need more information, like host OS. It might also be useful if you're able to git-bisect to find the exact commit that introduced this error

**Comment by aol-nnov** [comment]
@mwhooker Yes, ks file is processed correctly and OS installation succeeded. Moreover, while provisioning stage is hanging, I can ssh into vm with port printed in the packer logs.
Also it does not matter if you run it headless or not - same behaviour.

Actually, it's not pure centos on my side, but a "closed source clone" of it, so it doesn't really matter that you changed scripts to point 6.8 release.
To my understanding the problem resides on the host side, not on the vm side as it vanishes away if I barely ssh into the host before launching packer.

**Comment by SwampDragons** [comment]
A few months ago I made a change to how our communicator returns stdout output... if it really does have to do with large amounts of stdout text, it might be that this is fixed in 1.0.4.  OR, it might be that now that we're properly returning stdout we might find the actual error message.  Since we can't repro, I'm going to close but please reopen if this issues still exists with packer > 1.0.4
